cs3api4lab/api/cs3_file_api.py

```diff
@@ -261,6 +261,64 @@ def create_directory(self, path, endpoint=None):
         self.log.debug(
             'msg="Invoked create container" filepath="%s" elapsedTimems="%.1f"' % (path, (tend - tstart) * 1000))
 
+    def list_checkpoints(self, path):
+        print('list path', path)
+        ref = FileUtils.get_reference(path)
+        request =cs3sp.ListFileVersionsRequest(ref=ref)
+        response = self.cs3_api.ListFileVersions(request=request, metadata=[('x-access-token', self.auth.authenticate())])
+
+        print('list response', response)
+        if response.status.code != cs3code.CODE_OK:
+            self.log.error(f"msg=\"Failed to list file versions\" filepath={path}")
+
+        checkpoints = []
+        for i in reversed(range(len(response.versions)-1)):
+            if self.config.dev_env: # we need this until https://github.com/cs3org/reva/issues/3928 is resolved
+                last_modified = datetime.datetime.fromtimestamp(response.versions[i].mtime / 1000.0)
+            else:
+                last_modified = datetime.datetime.fromtimestamp(response.versions[i].mtime)
+
+            # because the checkpoint url is parsed by regular expression, so we need to replace the '.' sign from key
+            key = str(response.versions[i].key).replace('.', '-')
+
+            checkpoints.append({
+                'id': key,
+                'last_modified': last_modified
+            })
+
+        return checkpoints
+
+    def restore_checkpoint(self, checkpoint_id, path):
+        # because the checkpoint url is parsed by regular expression, so we need to replace the '.' sign from key
+        checkpoint_id = str(checkpoint_id).replace('-', '.')
+        if self.config.dev_env: # we need this until https://github.com/cs3org/reva/issues/3927 is fixed
+            checkpoint_id = 'v' + checkpoint_id
+
+        ref = FileUtils.get_reference(path, self.config.endpoint)
+        request = cs3sp.RestoreFileVersionRequest(ref=ref, key=checkpoint_id)
+
+        file_info = self.stat_info(path)
+        # additional request until this issue is resolved https://github.com/cs3org/reva/issues/3243
+        if self.config.dev_env and "/home/" in file_info['filepath']:
+            opaque_id = urllib.parse.unquote(file_info['inode']['opaque_id'])
+            storage_id = urllib.parse.unquote(file_info['inode']['storage_id'])
+            file_info = self.stat_info(opaque_id, storage_id)
+
+        if self.config.locks_api == "metadata":
+            if self.lock_api.is_file_locked(file_info):
+                self.log.error(f"File {path} is locked and cannot be restored")
+                raise IOError(f"File {path} is locked and cannot be restored")
+        elif self.config.locks_api == "cs3":
+            if self.lock_api.is_file_locked(file_info):
```

Comment:
If file is locked, it means that a lock exists and it's not ours, right?
So, the call bellow will always fail, no? Because we should never be able to restore if we're not locking it ourselves (or no lock exists).

Shouldn't we pass the lock id to the restore, only if there's a lock and it's ours?

Reply:
Given the comment above, maybe we can simplify this logic a bit.

Reply:
> Shouldn't we pass the lock id to the restore, only if there's a lock and it's ours?

This is precisely what we are doing, only if the file is locked and it's not ours where are passing the lock id to the request, the if statement determines if the file is locked and forms additional info (lock id) for the request object. The actual restore call is not in if-else statement. So if there's no lock we're using basic request object formed above and it's used for  "metadata" locks as well as "cs3" locks (when the file is not locked).